velox/substrait/VeloxToSubstraitPlan.cpp

```diff
@@ -170,13 +176,144 @@ void VeloxToSubstraitPlanConvertor::toSubstrait(
+            inputType));
+  }
+
+  // AggregatesSize should be equal to or greater than the aggregateMasks Size.
+  // Two cases: 1. aggregateMasksSize = 0, aggregatesSize > aggregateMasksSize.
+  // 2. aggregateMasksSize != 0, aggregatesSize = aggregateMasksSize.
+  auto aggregates = aggregateNode->aggregates();
+  auto aggregateMasks = aggregateNode->aggregateMasks();
+  int64_t aggregatesSize = aggregates.size();
+  int64_t aggregateMasksSize = aggregateMasks.size();
+  VELOX_CHECK_GE(aggregatesSize, aggregateMasksSize);
+
+  for (int64_t i = 0; i < aggregatesSize; i++) {
+    ::substrait::AggregateRel_Measure* aggMeasures =
+        aggregateRel->add_measures();
+
+    std::shared_ptr<const core::FieldAccessTypedExpr> aggMaskExpr =
+        aggregateMasks.at(i);
+    // Set substrait filter.
+    ::substrait::Expression* aggFilter = aggMeasures->mutable_filter();
+    if (aggMaskExpr.get()) {
+      aggFilter->MergeFrom(exprConvertor_->toSubstraitExpr(
+          arena,
+          std::dynamic_pointer_cast<const ITypedExpr>(aggMaskExpr),
+          inputType));
+    } else {
+      // Set null.
+      aggFilter = NULL;
+    }
+
+    // Process measure, eg:sum(a).
+    std::shared_ptr<const core::CallTypedExpr> aggregatesExpr =
+        aggregates.at(i);
+    ::substrait::AggregateFunction* aggFunction =
+        aggMeasures->mutable_measure();
+
+    // Aggregation function name.
+    std::string funName = aggregatesExpr->name();
+    // set aggFunction args.
+    std::vector<core::TypedExprPtr> aggregatesExprInputs =
+        aggregatesExpr->inputs();
+
+    int64_t aggregatesExprInputsSize = aggregatesExprInputs.size();
+    for (int64_t j = 0; j < aggregatesExprInputsSize; j++) {
+      ::substrait::Expression* aggFunctionExpr = aggFunction->add_args();
+      core::TypedExprPtr expr = aggregatesExprInputs[j];
+      // In velox plan, expr can only be the FieldAccessTypedExpr.
+      // If the expr is not FieldAccess, people need do project firstly.
+      if (auto vAggregatesExprInput =
+              std::dynamic_pointer_cast<const core::CallTypedExpr>(expr)) {
+        VELOX_NYI("In Velox Plan, the aggregates type cannot be CallTypedExpr");
+      } else if (
+          auto vAggregatesExprInput =
+              std::dynamic_pointer_cast<const core::ConstantTypedExpr>(expr) &&
+              funName == "count") {
+        // For case count(1).
+        aggFunction->clear_args();
+      } else {
+        // FieldAccessTypedExpr and Cast.
+        aggFunctionExpr->MergeFrom(
+            exprConvertor_->toSubstraitExpr(arena, expr, inputType));
+      }
+    }
+
+    // Set substrait aggregate Function reference and output type.
+    aggFunction->set_function_reference(functionMap_[funName]);
+
+    aggFunction->mutable_output_type()->MergeFrom(
+        typeConvertor_->toSubstraitType(arena, aggregatesExpr->type()));
+
+    // Set substrait aggregate Function phase.
+    switch (aggregateNode->step()) {
```

Comment:
consider extracting a helper method for readability

Reply:
done.